rllib/env/testing/single_agent_gym_env_runner.py

```diff
@@ -0,0 +1,143 @@
+        num_episodes: Optional[int] = None,
+        explore: bool = True,
+        random_actions: bool = False,
+        force_reset: bool = False,
+        **kwargs,
+    ) -> Tuple[List[Episode], List[Episode]]:
+        # So far, only sampling by num_timesteps is supported.
+        assert num_episodes is None
+
+        # If nothing provided, use our configured default settings.
+        if num_timesteps is None and num_episodes is None:
+            num_timesteps = self.config.rollout_fragment_length * self.num_envs
+
+        return self._sample_timesteps(
+            num_timesteps=num_timesteps,
+            explore=explore,
+            random_actions=random_actions,
+            force_reset=force_reset,
+        )
+
+    def _sample_timesteps(
+        self,
+        num_timesteps: int,
+        explore: bool = True,
+        random_actions: bool = False,
+        force_reset: bool = False,
+    ):
+        """Runs n timesteps on the environment(s) and returns experiences.
+
+        Timesteps are counted in total (across all vectorized sub-environments). For
+        example, if self.num_envs=2 and num_timesteps=10, each sub-environment
+        will be sampled for 5 steps.
+        """
+        done_episodes_to_return = []
+
+        # Have to reset the env (on all vector sub-envs).
+        if force_reset or self._needs_initial_reset:
+            obs, _ = self.env.reset()
+            # Start new episodes.
+            self._episodes = [Episode() for _ in range(self.num_envs)]
+            self._needs_initial_reset = False
+            # Set initial observations of the new episodes.
+            for i, o in enumerate(self._split_by_env(obs)):
+                self._episodes[i].add_initial_observation(initial_observation=o)
+
+        # Loop for as long as we have not reached `num_timesteps` timesteps.
+        ts = 0
+        while ts < num_timesteps:
+            # Act randomly.
+            actions = self.env.action_space.sample()
+            obs, rewards, terminateds, truncateds, infos = self.env.step(actions)
+            # Count timesteps across all environments.
+            ts += self.num_envs
+
+            for i, (o, a, r, term, trunc) in enumerate(
+                zip(
+                    self._split_by_env(obs),
+                    self._split_by_env(actions),
+                    self._split_by_env(rewards),
+                    self._split_by_env(terminateds),
+                    self._split_by_env(truncateds),
+                )
+            ):
+                # Episode is done (terminated or truncated).
+                if term or trunc:
+                    # Finish the episode object with the actual terminal observation
+                    # stored in the info dict (`o` is already the reset obs of the new
+                    # episode).
+                    self._episodes[i].add_timestep(
+                        infos["final_observation"][i], a, r, is_terminated=True
```

Comment:
do all gymnasium envs follow this convention now? o returned by `.step()` is the reset of the next env? Can we comment that detail here?

Reply:
So we should not necessarily hit `.reset()` anymore?

Reply:
Correct, with gymnasium.vector, you'll only have to reset once at the very beginning. After that, each sub-env resets itself automatically. Actually, there is a setting where you can switch off that behavior, but by default, it's on.
I'll add more explanations.

Reply:
This makes things much simpler now as e.g. in our current RLlib VectorEnv, we have to check individual environments for being terminated and then reset them individually.